pkg/mcs/resourcemanager/server/keyspace_manager.go

```diff
@@ -346,3 +347,127 @@ func (rt *ruTracker) getRUPerSec() float64 {
+	if serviceLimiter == nil || serviceLimiter.ServiceLimit == 0 {
+		return
+	}
+	priorityQueues := krgm.getPriorityQueues()
+	if len(priorityQueues) == 0 {
+		return
+	}
+	remainingServiceLimit := serviceLimiter.ServiceLimit
+	for _, queue := range priorityQueues {
+		if len(queue) == 0 {
+			continue
+		}
+		// If the remaining service limit is 0, set the fill rate of all the remaining resource groups to 0 directly.
+		if remainingServiceLimit == 0 {
+			for _, group := range queue {
+				ruTracker := krgm.getRUTracker(group.Name)
+				if ruTracker == nil || ruTracker.getRUPerSec() == 0 {
+					continue
+				}
+				// Only set the active resource groups.
+				group.overrideFillRate(0)
+			}
+			continue
+		}
+		// Calculate the total real-time RU demand of all resource groups in the current priority.
+		totalRUDemand := 0.0
+		ruDemandMap := make(map[string]float64, len(queue))
+		for _, group := range queue {
+			ruTracker := krgm.getRUTracker(group.Name)
+			// Not found the RU tracker, skip this group.
+			if ruTracker == nil {
+				continue
+			}
+			// The RU demand exceeding the fill rate setting should be ignored.
+			ruDemand := math.Min(ruTracker.getRUPerSec(), group.getFillRateSetting())
+			totalRUDemand += ruDemand
+			ruDemandMap[group.Name] = ruDemand
+		}
+		if totalRUDemand > remainingServiceLimit {
+			// If the total real-time RU demand is greater than the service limit,
+			// allocate the service limit to all resource groups based on the proportion of RU demand.
+			for _, group := range queue {
+				ruDemand := ruDemandMap[group.Name]
+				overrideFillRate := math.Min(
+					remainingServiceLimit*ruDemand/totalRUDemand,
+					group.getFillRateSetting(), // Should not exceed the fill rate setting.
+				)
+				group.overrideFillRate(overrideFillRate)
+			}
+			remainingServiceLimit = 0
+		} else {
+			// If the total real-time RU demand is less or equal to the service limit,
+			// allocate the needed RU to the resource group respectively.
+			for _, group := range queue {
+				// By setting the override fill rate to -1 to allow the resource group to consume
+				// as many RUs as they originally need according to its fill rate setting. This is
+				// to prevent the true demand from being suppressed due to setting `overrideFillRate`,
+				// thereby wasting idle resources of the Service Limit.
+				group.overrideFillRate(-1)
+			}
+			// Although this layer does not set `overrideFillRate`, it still needs to deduct the actual
+			// RU consumption from `remainingFillRate` to reflect the concept of priority, so that the
+			// available service limit share decreases level by level.
+			remainingServiceLimit -= totalRUDemand
+		}
+	}
+}
+
+func (krgm *keyspaceResourceGroupManager) getPriorityQueues() [][]*ResourceGroup {
+	priorityQueues := make([][]*ResourceGroup, maxPriority+1)
+	krgm.RLock()
+	for _, group := range krgm.groups {
+		if group.Name == DefaultResourceGroupName {
```

Comment:
Why bypass the default group? It should also be controlled especially when there are high priority groups.